xds/internal/balancer/clusterimpl/clusterimpl.go

```diff
@@ -305,6 +310,34 @@ func (cib *clusterImplBalancer) UpdateState(state balancer.State) {
 	cib.pickerUpdateCh.Put(state)
 }
 
+func (cib *clusterImplBalancer) getClusterName() (string, bool) {
+	v := cib.clusterName.Load()
+	vv, ok := v.(string)
+	return vv, ok
+}
+
+func (cib *clusterImplBalancer) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if clusterName, ok := cib.getClusterName(); ok {
+		newAddrs := make([]resolver.Address, len(addrs))
+		for i, addr := range addrs {
+			newAddrs[i] = xdsinternal.SetHandshakeClusterName(addr, clusterName)
+		}
+		addrs = newAddrs
+	}
+	return cib.ClientConn.NewSubConn(addrs, opts)
+}
+
+func (cib *clusterImplBalancer) UpdateAddresses(sc balancer.SubConn, addrs []resolver.Address) {
+	if clusterName, ok := cib.getClusterName(); ok {
+		newAddrs := make([]resolver.Address, len(addrs))
+		for i, addr := range addrs {
+			newAddrs[i] = xdsinternal.SetHandshakeClusterName(addr, clusterName)
+		}
+		addrs = newAddrs
+	}
+	cib.ClientConn.UpdateAddresses(sc, addrs)
```

Comment:
Same question here as in `NewSubConn`.

Reply:
Maybe a comment pointing to the comment in `NewSubConn()` which goes over the options when clusterName is missing?

Reply:
No longer needed with mutex.